File: src/main/java/EngimonHunter2000/Player.java
package EngimonHunter2000;

import java.io.Serializable;

/**
 * class untuk menyimpan semua keperluan Player {@link EngimonHunter2000}
 * @author Cynthia Rusadi
 */

public class Player implements Serializable {
    public static final long serialVersionUID = 1L;
    private Inventory<Engimon> listEngimon;
    private Inventory<Item> listItem;
    private Engimon activeEngi;
    private Position position;
    private char dir;

    // constructor
    public Player(EngiDex dex)
            throws EngimonSpeciesException, ElementsListException, EngimonException, InventoryException {
        Engimon engie = new Engimon(dex, "Picakhu", "Picakhu");
        engie.addLevel(10);
        Inventory.setMax(50);
        this.listEngimon = new Inventory<Engimon>();
        this.listEngimon.addItem(engie);
        this.listItem = new Inventory<Item>();
        this.activeEngi = engie;
        this.dir = 'a';
        this.position = new Position();
    }

    // getters
    public Inventory<Engimon> getInventoryEngimon() {
        return this.listEngimon;
    }

    public Inventory<Item> getInventoryItem() {
        return this.listItem;
    }

    public Engimon getActiveEngimon() {
        return this.activeEngi;
    }

    public Position getPosition() {
        return this.position;
    }

    public int getPositionX() {
        return this.position.getX();
    }

    public int getPositionY() {
        return this.position.getY();
    }

    public char getDir() {
        return this.dir;
    }

    // setters
    public void setPosition(Position pos) {
        this.position.setPosition(pos);
    }

    public void setPositionX(int pos) {
        this.position.setX(pos);
    }

    public void setPositionY(int pos) {
        this.position.setY(pos);
    }

    public void setDir(char _dir) {
        this.dir = _dir;
    }

    // methods
    public void checkActiveEngimon() {
        this.activeEngi.showEngimon();
    }

    public void switchEngimon(int engiIdx) throws InventoryException {
        Position currPos = this.activeEngi.getPosition();
        this.activeEngi.setPos(-1, -1);

        this.listEngimon.at(engiIdx).setPos(currPos.getX(), currPos.getY());
        this.activeEngi = this.listEngimon.at(engiIdx);
    }

    // specifically only one engimon
    public void showEngimon(int engiIdx) throws InventoryException {
        this.listEngimon.at(engiIdx).showEngimon();
    }

    // print all engimon in inventory
    public void showEngimon() throws InventoryException {
        if (this.listEngimon.getAllInvenTotalItemCount() == 0)
            throw new InventoryException(1);
        System.out.println("Engimon di dalam Inventory:");
        this.listEngimon.showInventory();
    }

    public void changeEngimonName(int engiIdx, String newName) throws EngimonException, InventoryException {
        this.listEngimon.at(engiIdx).setName(newName);
    }

    public void addEngimon(Engimon engi) throws InventoryException {
        this.listEngimon.addItem(engi);
    }

    public void useItem(int itemIdx) throws ItemException, InventoryException, SkillEngimonException {
        int engiIdx = this.listEngimon.getItemFromIdx(this.activeEngi);
        if (engiIdx != -1) {
            this.listItem.at(itemIdx).learn(this.listEngimon.at(engiIdx));
            int qty = this.listItem.at(itemIdx).getQuantity();
            if (qty - 1 == 0) {
                listItem.removeItem(listItem.at(itemIdx));
            } else {
                this.listItem.at(itemIdx).setQuantity(qty - 1);
            }
        } else {
            throw new InventoryException(3);
        }
    }

    public void addItem(Item item) throws InventoryException {
        this.listItem.addItemNoDupe(item);
    }

    public void removeEngimon(int idx) throws InventoryException {
        this.listEngimon.removeItem(this.listEngimon.at(idx));
        if (this.listEngimon.at(idx).equals(this.activeEngi)) {
            this.activeEngi = this.listEngimon.at(0);
            // kalo broken, artinya dia cuman punya 1 engimon
        }
    }

    public void removeActiveEngimon() throws InventoryException {
        int idx = this.listEngimon.getItemFromIdx(this.activeEngi);
        if (idx != -1) {
            this.listEngimon.removeItem(this.listEngimon.at(idx));
        } else {
            throw new InventoryException(3);
        }
    }

    public void removeItem(int itemIdx, int n) throws InventoryException {
        if (this.listItem.at(itemIdx).getQuantity() == n) {
            this.listItem.removeItem(listItem.at(itemIdx), n);
        } else {
            for (int i = 0; i < n; i++) {
                this.listItem.at(itemIdx).decQuantity();
            }
            this.listItem.removeItemNoDupe(this.listItem.at(itemIdx), n);
        }
    }

    public void interact() {
        this.activeEngi.interact();
    }

    /**
     * Fungsi untuk mendapatkan engimon player yang memiliki level tertiinggi
     */
    public int getHighestEngimonLevel() {
        int highest = 0xFFFFFFFF;
        for (int i = 0; i < listEngimon.getAllInvenTotalItemCount(); ++i) {
            try {
                Engimon disEngie = listEngimon.at(i);
                highest = disEngie.getLvl() > highest ? disEngie.getLvl() : highest;
            } catch (InventoryException e) {
                System.err.println(e.what());
                break; // idk what to do, really ¯\_(ツ)_/¯
            }
        }

        return highest;
    }
}

File: src/impl/Player.cpp
#include "../headers/Player.hpp"

Player::Player()
    : name("yee"),
      // constructor objek lain udah immplied
      position(make_tuple(1, 1)),
      dir('a') {
    addEngimon(activeEngi);
}

string Player::getName() const { return name; }
Engimon& Player::getActiveEngimon() { return activeEngi; }

tuple<int, int> Player::getPosition() const { return position; }
int Player::getPositionX() const { return get<0>(position); }
int Player::getPositionY() const { return get<1>(position); }

char Player::getDir() const { return dir; }
Engimon& Player::getEngimonById(int idx){
    return listEngimon[idx];
}
Engimon& Player::getEngimonFromName(string engiName) {
    if (listEngimon.getItemCount() == 0) throw InventoryException(1);
    for (int i = 0; i < listEngimon.getItemCount(); i++) {
        if (engiName.compare(listEngimon[i].getName()) == 0) {
            return listEngimon[i];
        }
    }
    throw PlayerException(0);
}

Item& Player::getItemFromName(string _itemName) {
    if (listItem.getItemCount() == 0) throw InventoryException(1);
    for (int i = 0; i < listItem.getItemCount(); i++) {
        if (_itemName.compare(listItem[i].getName()) == 0) {
            return listItem[i];
        }
    }
    throw PlayerException(1);
}

int Player::getEngimonIdxFromName(string engiName) {
    Engimon temp = getEngimonFromName(engiName);
    return listEngimon.getFirstItemIndex(temp);
}

int Player::getItemIdxFromName(string _itemName) {
    Item temp = getItemFromName(_itemName);
    return listItem.getFirstItemIndex(temp);
}

void Player::setName(string _name) { name = _name; }

void Player::setPosition(tuple<int, int> pos) { position = pos; }
void Player::setPositionX(int x) { get<0>(position) = x; }
void Player::setPositionY(int y) { get<1>(position) = y; }

void Player::setDir(char _dir) { dir = _dir; }

void Player::checkActiveEngimon() { activeEngi.showEngimon(); }

void Player::switchEngimon(int engi) {
    int beforeEngi = listEngimon.getFirstItemIndex(activeEngi);
    listEngimon[beforeEngi].setPos(-1, -1);

    tuple<int, int> currPos = activeEngi.getPosition();
    listEngimon[engi].setPos(get<0>(currPos), get<1>(currPos));

    activeEngi = listEngimon[engi];
}

void Player::showEngimon(int idx) { listEngimon[idx].showEngimon(); }
void Player::showEngimon() const {
    cout << "Engimon di dalam Inventory: " << endl;
    listEngimon.showInventory();
}

void Player::addEngimon(Engimon engi) { listEngimon.addItem(engi); }
void Player::addEngimon(string engi) {
    Engimon temp = getEngimonFromName(engi);
    addEngimon(temp);
}

void Player::removeEngimon(Engimon engi) { listEngimon.removeItem(engi); }
void Player::removeEngimon(string engi) {
    Engimon temp = getEngimonFromName(engi);
    removeEngimon(temp);
}

void Player::showItem(string _item) {
    Item temp = getItemFromName(_item);
    temp.showItem();
}
void Player::showItem(int _item) { listItem[_item].showItem(); }
void Player::showItem() const {
    cout << "Skill Item di dalam Inventory: " << endl;
    listItem.showInventory();
}

void Player::useItem(int engi, int _item, const Dex& dex) {
    listItem[_item].learn(listEngimon[engi], dex);
    if (listItem[_item].getQuantity() == 0) {
        removeItem(listItem[_item]);
    }
}

void Player::addItem(Item _item) { listItem.addItemNoDupe(_item); }
void Player::addItem(string _item) {
    Item temp = getItemFromName(_item);
    addItem(temp);
}

void Player::removeItem(Item _item) { listItem.removeItemNoDupe(_item); }
void Player::removeItem(string _item) {
    Item temp = getItemFromName(_item);
    removeItem(temp);
}

void Player::interact() const { activeEngi.interact(); }

void Player::engimonIsEmpty() {
    if (listEngimon.getItemCount() == 0) throw InventoryException(1);
}
void Player::itemIsEmpty() {
    if (listItem.getItemCount() == 0) throw InventoryException(1);
}
void Player::inventoryIsFull() {
    if (!listItem.isFull()) throw InventoryException(0);
}

// exception
PlayerException::PlayerException(int x) : msgID(x) {}
const char* PlayerException::what() { return msg[msgID].c_str(); }
void PlayerException::bruh() { cout << what() << endl; }
string PlayerException::msg[] = {
    "Tidak ada Engimon tersebut di dalam Inventory",
    "Tidak ada Item Skill tersebut di dalam Inventory"};
